src/DeviceDataStreamValues.tsx

```diff
@@ -0,0 +1,219 @@
+  if (interfaceData.length === 0) {
+    return <p>No data in the selected timeframe.</p>;
+  } else if (aggregation === 'individual') {
+    return <IndividualDatastreamTable data={interfaceData as AstarteDatastreamIndividualData[]} />;
+  }
+  return <ObjectDatastreamTable data={interfaceData as AstarteDatastreamObjectData[]} />;
+};
+
+export default (): React.ReactElement => {
+  const { interfaceName = '', deviceId = '', interfaceMajor } = useParams();
+  const astarte = useAstarte();
+  const [requestingData, setRequestingData] = useState(false);
+  const [fetchedData, setFetchedData] = useState<AstarteInterfaceValues | null>(null);
+  const [formAlerts, formAlertsController] = useAlerts();
+  const [interfaceData, setInterfaceData] = useState<AstarteInterface | null>(null);
+
+  const deviceDataFetcher = useFetch(() =>
+    astarte.client.getDeviceData({
+      deviceId,
+      interfaceName,
+    }),
+  );
+
+  const fetchData = async (path?: string, since?: string, to?: string) => {
+    setRequestingData(true);
+    setFetchedData(null);
+    astarte.client
+      .getDeviceData({
+        deviceId,
+        interfaceName,
+        path,
+        since,
+        to,
+      })
+      .then((data) => {
+        setFetchedData(data);
+      })
+      .catch((err) => {
+        formAlertsController.showError(
+          `Could not fetch data to interface: ${err.response.data.errors.detail}`,
+        );
+        setFetchedData([]);
+      })
+      .finally(() => {
+        setRequestingData(false);
+      });
+  };
+
+  useEffect(() => {
+    if (interfaceMajor) {
+      const major = parseInt(interfaceMajor, 10);
+      astarte.client
+        .getInterface({ interfaceName, interfaceMajor: major })
+        .then((data) => setInterfaceData(data));
+    }
+  }, []);
+
+  return (
+    <Container fluid className="p-3">
+      <div className="d-flex justify-content-between">
+        <h2>
+          <BackButton href={`/devices/${deviceId}/interfaces/${interfaceName}/${interfaceMajor}`} />
+          Interface Datastream Data
+        </h2>
+      </div>
+      <AlertsBanner alerts={formAlerts} />
+      <Card className="mt-4">
+        <Card.Header>
+          <span className="font-monospace">{deviceId}</span> /{interfaceName}
+        </Card.Header>
+        <Card.Body>
+          <FilterDatastreamData
+            data={deviceDataFetcher.value}
+            fetchData={fetchData}
+            requestingData={requestingData}
+          />
+          {fetchedData && deviceDataFetcher.value && (
+            <RenderDatastreamData
+              data={fetchedData}
+              interfaceData={interfaceData}
+              values={deviceDataFetcher.value}
+            >
+              {(transformedData) => (
+                <InterfaceData
+                  interfaceData={transformedData}
+                  aggregation={interfaceData?.aggregation}
+                />
+              )}
+            </RenderDatastreamData>
```

Comment:
The `RenderDatastreamData` component does not render any UI or data, in contrast to its name: it's only used to call `transformObjectData` / `transformIndividualData` from `transformData.ts`.

Since the functions contained within `transformData.ts` are meant to manipulate data for the scope of rendering it here, they are not really generic or reusable; I would move this functions here in the component, since they are tightly coupled with it, and remove the `transformData.ts` file.

We can also avoid to define the `RenderDatastreamData` component, and just derive a new variable from `fetchedData` using `transformObjectData` / `transformIndividualData` before rendering data.